contracts/FeeDisbursal/FeeDisbursal.sol

```diff
@@ -0,0 +1,269 @@
+            proRataToken: proRataToken,
+            snapshotId: snapshotId,
+            disburser: wallet
+        }));
+        //@TODO: add log message
+    }
+
+    // implementation of reclaimbale
+    function reclaim(IBasicToken token)
+        public
+        only(ROLE_RECLAIMER)
+    {   
+        // forbid reclaiming any or our platform payment tokens
+        // @TODO: move access control to Controller
+        require(!isDisbursableToken(address(token)));
+        Reclaimable.reclaim(token);
+    }
+
+    /// @notice helper to determine if the token at the given address is supported for disbursing
+    /// @param token address of token in question
+    function isDisbursableToken(address token)
+        public
+        view
+        returns (bool)
+    {   
+        // @TODO: migrate this to new, more flexible token registering in universe
+        if (token == address(UNIVERSE.etherToken())) return true;
+        if (token == address(UNIVERSE.euroToken())) return true;
+        return false;
+    }
+
+    ////////////////////////
+    // Private functions
+    ////////////////////////
+
+    /// @notice claim a token for an spender, returns the amount of tokens claimed
+    /// @param token address of the claimable token
+    /// @param spender address of the spender that will receive the funds
+    /// @param until until what index to claim to
+    function claimPrivate(address token, address spender, uint256 until)
+    public
+    returns (uint256 claimedAmount, uint256 lastIndex)
+    {
+        require(isDisbursableToken(token), "NF_DISB_UKNOWN_TOKEN");
+        (claimedAmount, lastIndex) = claimablePrivate(token, spender, until);
+
+        // mark spender disbursal progress
+        _disbursalProgress[token][spender] = lastIndex;
+
+        // do the actual token transfer
+        IERC223Token ierc223Token = IERC223Token(token);
+        ierc223Token.transfer(spender, claimedAmount, "");
+        //@TODO: add log message
+    }
+
+    /// @notice get the amount of tokens that can be claimed by a given spender
+    /// @param token address of the claimable token
+    /// @param spender address of the spender that will receive the funds
+    /// @param until until what index to claim to, use UINT256_MAX for all
+    function claimablePrivate(address token, address spender, uint256 until)
+    public
+    view
+    returns (uint256 claimableAmount, uint256 lastIndex)
+    {
+        require(isDisbursableToken(token), "NF_DISB_UKNOWN_TOKEN");
+        lastIndex = min(until, _disbursals[token].length);
+        claimableAmount = 0;
+        for (uint256 i = _disbursalProgress[token][spender]; i < lastIndex; i += 1) {
+            Disbursal storage disbursal = _disbursals[token][i];
+            BasicSnapshotToken proRataToken = BasicSnapshotToken(disbursal.proRataToken);
+            uint256 snapshotId = disbursal.snapshotId;
+            uint256 proRataTotalSupply = proRataToken.totalSupplyAt(snapshotId);
```

Comment:
please store this in Disbursal struct! call is 900 unit of gas then you need to access storage in internal struct of the token. so it's better to cache it. it's 200 gas units per slot read. and we are shaving off gas.
so it makes sense to cache it for the good of the community. first claimer would pay for this obviously but we can assume that he's going to be a rich guy

Reply:
Done, good point!

Reply:
Actually I though about it. If we have unsealed disbursals, storing this will not work properly, as we might have a mismatch of totalSupply at the point of claiming, because there could always be more prorata tokens minted after the disbursal was made, no?